### PR TITLE
Upgrade to Spock 2.1

### PR DESCRIPTION
**Pull request by lptr**
Some commits extracted from https://github.com/gradle/gradle/pull/18552 that make sense even with Groovy 3 and Spock 2.0.

### CODE REVIEW

**Comment by lptr** [comment]
@bot-gradle test this please.

**Comment by lptr** [comment]
@bot-gradle test this please.

**Comment by lptr** [comment]
@bot-gradle test this please.

**Comment by lptr** [comment]
@bot-gradle test this please.

File: subprojects/model-core/src/testFixtures/groovy/org/gradle/api/internal/provider/ProviderTestUtil.java [hunk 1]
```diff
@@ -24,6 +24,7 @@
 import java.util.Iterator;
 import java.util.List;
 
+@SuppressWarnings("unchecked")
```
Comment (wolfs) [inline subprojects/model-core/src/testFixtures/groovy/org/gradle/api/internal/provider/ProviderTestUtil.java:27]: Which warning is being suppressed here?

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/rules/ComponentMetadataRulesSupport.groovy [hunk 2]
```diff
@@ -1,39 +0,0 @@
...
- * limitations under the License.
- */
-
-package org.gradle.integtests.resolve.rules
-
-import groovy.transform.CompileStatic
-import org.gradle.api.internal.artifacts.ivyservice.NamespaceId
-
-@CompileStatic
-trait ComponentMetadataRulesSupport {
```
Comment (wolfs) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/rules/ComponentMetadataRulesSupport.groovy:23]: In the commit you said that
```
Was only used in IvySpecificComponentMetadataRulesIntegrationTest, and it introduced some weird behavior with Groovy 4.
```

What was the weird behaviour? Is there a Groovy issue for the weird behaviour?

File: subprojects/base-services/build.gradle.kts [hunk 3]
```diff
@@ -14,6 +14,13 @@ tasks.named<JavaCompile>("compileTestJava") {
     options.release.set(8)
 }
 
+/**
+ * Use Java 8 compatibility for JMH benchmarks
+ */
+tasks.named<JavaCompile>("jmhCompileGeneratedClasses") {
+    options.release.set(8)
```
Comment (wolfs) [inline subprojects/base-services/build.gradle.kts:21]: Should this go to the `gradlebuild.jmh` plugin? Any idea why it doesn't compile with Spock 2.2?

File: subprojects/language-java/build.gradle.kts [hunk 4]
```diff
@@ -61,6 +61,8 @@ dependencies {
         because("ProjectBuilder test (JavaLanguagePluginTest) loads services from a Gradle distribution.")
     }
 
+    // TODO: Make these available for all integration tests? Maybe all tests?
```
Comment (wolfs) [inline subprojects/language-java/build.gradle.kts:64]: Do you have an answer for this todo? Why do we need the jetbrains annotations in `languageJava`?

File: subprojects/soak/build.gradle.kts [hunk 5]
```diff
@@ -18,6 +18,7 @@ dependencies {
     integTestImplementation(project(":persistent-cache"))
     integTestImplementation(libs.slf4jApi)
     integTestImplementation(libs.jetty)
+    integTestImplementation(libs.assertj)
```
Comment (wolfs) [inline subprojects/soak/build.gradle.kts:21]: Why do we need assertj in the soak tests? Looks like we didn't need it before, right?

**Review by wolfs — COMMENTED** [review]
I left a bunch of questions.

File: subprojects/language-java/build.gradle.kts (hunk 4 above)
Comment (lptr) [reply subprojects/language-java/build.gradle.kts:64]: `@Language` comes from them.

I don't have a good answer. In general I am against using the JetBrains annotations widely, because they introduce their own `@Nullable`/`@NotNull` etc. annotations. But then again, having them present in only some subprojects is also problematic. So no, I have no answer.

The intention here is to make the build work with Spock 2.2, though, that does not include the transitive dependency on JetBrains annotations anymore.

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/rules/ComponentMetadataRulesSupport.groovy (hunk 2 above)
Comment (lptr) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/rules/ComponentMetadataRulesSupport.groovy:23]: I think it was this: https://issues.apache.org/jira/browse/GROOVY-10466

File: subprojects/model-core/src/testFixtures/groovy/org/gradle/api/internal/provider/ProviderTestUtil.java (hunk 1 above)
Comment (lptr) [reply subprojects/model-core/src/testFixtures/groovy/org/gradle/api/internal/provider/ProviderTestUtil.java:27]: I guess it has something to do with the generic varargs.

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/rules/ComponentMetadataRulesSupport.groovy (hunk 2 above)
Comment (lptr) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/rules/ComponentMetadataRulesSupport.groovy:23]: Actually, that means we can drop that commit from here.

File: subprojects/model-core/src/testFixtures/groovy/org/gradle/api/internal/provider/ProviderTestUtil.java (hunk 1 above)
Comment (wolfs) [reply subprojects/model-core/src/testFixtures/groovy/org/gradle/api/internal/provider/ProviderTestUtil.java:27]: Can you use `@SafeVargs` to get rid of the deprecation message?

File: subprojects/language-java/build.gradle.kts (hunk 4 above)
Comment (wolfs) [reply subprojects/language-java/build.gradle.kts:64]: 👍

File: subprojects/model-core/src/testFixtures/groovy/org/gradle/api/internal/provider/ProviderTestUtil.java (hunk 1 above)
Comment (lptr) [reply subprojects/model-core/src/testFixtures/groovy/org/gradle/api/internal/provider/ProviderTestUtil.java:27]: I'm removing this as it's not a problem with Groovy 4 either.

File: subprojects/base-services/build.gradle.kts (hunk 3 above)
Comment (lptr) [reply subprojects/base-services/build.gradle.kts:21]: I think there is a problem here with how `usedInWorkers()` works, as it will override the setting set by the JMH plugin. That's why above we also set the target for `compileTestJava` the same way.

File: subprojects/soak/build.gradle.kts (hunk 5 above)
Comment (wolfs) [reply subprojects/soak/build.gradle.kts:21]: Let me rephrase my question: Why do soak tests use `assertj`? I am surprised by this since we have been using Spock for Soak tests as well.

File: subprojects/base-services/build.gradle.kts (hunk 3 above)
Comment (wolfs) [reply subprojects/base-services/build.gradle.kts:21]: Fine with me.

File: subprojects/soak/build.gradle.kts (hunk 5 above)
Comment (lptr) [reply subprojects/soak/build.gradle.kts:21]: I've added the actual upgrade to Spock 2.1 (which is essentially the same as 2.2-M1, but it has the Groovy 4 checks in place). Now we need this change.

**Comment by lptr** [comment]
@bot-gradle test this please.

File: subprojects/soak/build.gradle.kts (hunk 5 above)
Comment (lptr) [reply subprojects/soak/build.gradle.kts:21]: > Let me rephrase my question: Why do soak tests use `assertj`? I am surprised by this since we have been using Spock for Soak tests as well.

No idea, but let's investigate that separately, shall we?

**Comment by lptr** [comment]
@bot-gradle test this please.

**Comment by lptr** [comment]
@bot-gradle test this please.

**Comment by lptr** [comment]
@bot-gradle test this please.

File: subprojects/soak/build.gradle.kts (hunk 5 above)
Comment (wolfs) [reply subprojects/soak/build.gradle.kts:21]: I took a look. Can you add a
```
because "Kotlin Soak tests use AssertJ"
```

to the dependency?

File: subprojects/distributions-dependencies/build.gradle.kts [hunk 6]
```diff
@@ -103,9 +104,9 @@ dependencies {
         api(libs.jsr305)                { version { strictly("3.0.2") }}
         api(libs.julToSlf4j)            { version { strictly(slf4jVersion) }}
         api(libs.junit)                 { version { strictly("4.13.2") }}
-        api(libs.junit5JupiterApi)      { version { strictly("5.7.2") }}
-        api(libs.junit5Vintage)         { version { strictly("5.7.2") }}
-        api(libs.junitPlatform)         { version { strictly("1.7.2") }}
+        api(libs.junit5JupiterApi)      { version { strictly("5.8.1") }}
```
Comment (wolfs) [inline subprojects/distributions-dependencies/build.gradle.kts:107]: Can we use 5.8.2/1.8.2 here?

File: subprojects/distributions-integ-tests/src/integTest/groovy/org/gradle/DistributionIntegrationSpec.groovy [hunk 7]
```diff
@@ -57,7 +57,7 @@ abstract class DistributionIntegrationSpec extends AbstractIntegrationSpec {
      * Change this whenever you add or remove subprojects for distribution-packaged plugins (lib/plugins).
      */
     int getPackagedPluginsJarCount() {
-        46
+        45
```
Comment (wolfs) [inline subprojects/distributions-integ-tests/src/integTest/groovy/org/gradle/DistributionIntegrationSpec.groovy:60]: Which dependency has been removed here? Is removing the dependency a breaking change?

File: gradle/verification-metadata.xml [hunk 8]
```diff
@@ -1051,6 +1051,11 @@
             <pgp value="77993d3295698c46912e83984b2277c6358670ee"/>
          </artifact>
       </component>
+      <component group="org.assertj" name="assertj-core" version="3.20.2">
```
Comment (wolfs) [inline gradle/verification-metadata.xml:1054]: Is it possible to consolidate the assertj versions we use? I think it would be great to only use 3.20 (or whathever the newest version is) instead of three different versions.

**Review by wolfs — COMMENTED** [review]
I left some more comments.

File: subprojects/distributions-dependencies/build.gradle.kts [hunk 9]
```diff
@@ -106,7 +106,7 @@ dependencies {
         api(libs.junit)                 { version { strictly("4.13.2") }}
         api(libs.junit5JupiterApi)      { version { strictly("5.7.2") }}
         api(libs.junit5Vintage)         { version { strictly("5.7.2") }}
-        api(libs.junitPlatform)         { version { strictly("1.7.2") }}
+        api(libs.junitPlatform)         { version { strictly("1.8.1") }}
```
Comment (leonard84) [inline subprojects/distributions-dependencies/build.gradle.kts:109]: Why are you managing the versions separately here? Consider using the `junit-bom` to get properly aligned versions. The JUnit Platform version should match the Jupiter version if you ignore the major version.

Comment (lptr) [reply subprojects/distributions-dependencies/build.gradle.kts:109]: I don't know why we ended up with this setup. Could you give me a diff that'd make this more correct/simpler?

Comment (leonard84) [reply subprojects/distributions-dependencies/build.gradle.kts:109]: I couldn't figure out where the `libs` definitions come from, but I would assume you could to something like `api(platform(libs.junitPlatformBom)) {...}` for https://search.maven.org/artifact/org.junit/junit-bom/5.8.2/pom 

If it is possible to use `platform` in dependency constraints.

Comment (lptr) [reply subprojects/distributions-dependencies/build.gradle.kts:109]: Thanks. Let's do this in a separate PR.

File: gradle/verification-metadata.xml (hunk 8 above)
Comment (lptr) [reply gradle/verification-metadata.xml:1054]: It's `apiguardian-api.jar`. I guess it must have been dropped from JUnit?

File: subprojects/distributions-integ-tests/src/integTest/groovy/org/gradle/DistributionIntegrationSpec.groovy (hunk 7 above)
Comment (lptr) [reply subprojects/distributions-integ-tests/src/integTest/groovy/org/gradle/DistributionIntegrationSpec.groovy:60]: It's `apiguardian-api.jar`. I guess it must have been dropped from JUnit?

Comment (lptr) [reply subprojects/distributions-integ-tests/src/integTest/groovy/org/gradle/DistributionIntegrationSpec.groovy:60]: Yeah, it used to be a dependency of junitplatform 1.7.2:

```text
|    |    |    +--- project :testing-junit-platform
|    |    |    |    +--- project :base-services (*)
|    |    |    |    +--- project :messaging (*)
|    |    |    |    +--- project :platform-jvm (*)
|    |    |    |    +--- project :language-java (*)
|    |    |    |    +--- project :testing-base (*)
|    |    |    |    +--- project :testing-jvm (*)
|    |    |    |    \--- org.junit.platform:junit-platform-launcher -> 1.7.2
|    |    |    |         +--- org.junit:junit-bom:5.7.2
|    |    |    |         |    +--- org.junit.platform:junit-platform-engine:1.7.2 (c)
|    |    |    |         |    +--- org.junit.platform:junit-platform-launcher:1.7.2 (c)
|    |    |    |         |    \--- org.junit.platform:junit-platform-commons:1.7.2 (c)
|    |    |    |         +--- org.apiguardian:apiguardian-api:1.1.0
|    |    |    |         \--- org.junit.platform:junit-platform-engine:1.7.2
|    |    |    |              +--- org.junit:junit-bom:5.7.2 (*)
|    |    |    |              +--- org.apiguardian:apiguardian-api:1.1.0
|    |    |    |              +--- org.opentest4j:opentest4j:1.2.0
|    |    |    |              \--- org.junit.platform:junit-platform-commons:1.7.2
|    |    |    |                   +--- org.junit:junit-bom:5.7.2 (*)
|    |    |    |                   \--- org.apiguardian:apiguardian-api:1.1.0
```

https://ge.gradle.org/s/soxzctqhtkyyg/dependencies?dependencies=guar&expandAll&focusedDependency=WzI2LDAsMjI0NSxbMCwwLFswXV1d&focusedDependencyView=versions

File: gradle/verification-metadata.xml (hunk 8 above)
Comment (lptr) [reply gradle/verification-metadata.xml:1049]: I removed one that is not used, but this problem would be better solved by enabling `--write-verification-metadata` to remove unused entries. I believe @ljacomet will implement that at some point.

File: subprojects/soak/build.gradle.kts (hunk 5 above)
Comment (lptr) [reply subprojects/soak/build.gradle.kts]: Added.

**Review by wolfs — APPROVED** [review]
LGTM!

**Comment by lptr** [comment]
@bot-gradle test and merge please.
